packages/flutter/lib/src/material/data_table.dart

```diff
@@ -0,0 +1,757 @@
+  /// a checkbox that can be checked to select all selectable rows
+  /// (and which is checked if all the rows are selected), and each
+  /// subsequent row will have a checkbox to toggle just that row.
+  ///
+  /// Rows whose [onSelectChanged] callback is null are ignored for
+  /// the purposes of determining the state of the "all" checkbox;
+  /// their checkbox is disabled.
+  final ValueChanged<bool> onSelectChanged;
+
+  /// Whether the row is selected.
+  ///
+  /// If [onSelectChanged] is non-null for any row in the table, then
+  /// a checkbox is shown at the start of each row. If the row is
+  /// selected (true), the checkbox will be checked and the row will
+  /// be highlighted.
+  ///
+  /// Otherwise, the checkbox, if present, will not be checked.
+  final bool selected;
+
+  /// The data for this row.
+  ///
+  /// There must be exactly as many cells as there are columns in the
+  /// table.
+  final List<DataCell> cells;
+
+  bool get _debugInteractive => onSelectChanged != null || cells.any((DataCell cell) => cell._debugInteractive);
+}
+
+/// The data for a cell of a [DataTable].
+///
+/// One list of [DataCell] objects must be provided for each [DataRow]
+/// in the [DataTable], in the [new DataRow] constructor's `cells`
+/// argument.
+class DataCell {
+  /// Creates an object to hold the data for a cell in a [DataTable].
+  ///
+  /// The first argument is the widget to show for the cell, typically
+  /// a [Text] or [DropDownButton] widget; this becomes the [widget]
+  /// property and must not be null.
+  ///
+  /// If the cell has no data, then a [Text] widget with placeholder
+  /// text should be provided instead, and then the [placeholder]
+  /// argument should be set to true.
+  const DataCell(this.widget, {
+    this.placeholder: false,
+    this.showEditIcon: false,
+    this.onTap
+  });
+
+  /// The data for the row.
+  ///
+  /// Typically a [Text] widget or a [DropDownButton] widget.
+  ///
+  /// If the cell has no data, then a [Text] widget with placeholder
+  /// text should be provided instead, and [placeholder] should be set
+  /// to true.
+  final Widget widget;
+
+  /// Whether the [widget] is actually a placeholder.
+  ///
+  /// If this is true, the default text style for the cell is changed
+  /// to be appropriate for placeholder text.
+  final bool placeholder;
+
+  /// Whether to show an edit icon at the end of the cell.
+  ///
+  /// This does not make the cell actually editable; the caller must
+  /// implement editing behavior if desired (by hooking [onTap]).
+  ///
+  /// If this is set, [onTap] should also be set, otherwise tapping
+  /// the icon will have no effect.
+  final bool showEditIcon;
+
+  /// Invoked if the cell is tapped.
+  ///
+  /// If non-null, the cell will have an ink well.
+  final VoidCallback onTap;
+
+  bool get _debugInteractive => onTap != null;
+}
+
+/// A material design data table.
+///
+/// Displaying data in a table is expensive, because to lay out the
```

Comment:
Perhaps we should allow one to specify explicit widths for some columns and relative-width-flexibility values for the rest, as an alternative to measuring the column widths.